core/src/main/java/org/elasticsearch/ElasticsearchException.java

```diff
@@ -324,21 +335,74 @@ public static void toXContent(XContentBuilder builder, Params params, Throwable
         if (ex instanceof ElasticsearchException) {
             ((ElasticsearchException) ex).toXContent(builder, params);
         } else {
-            builder.field("type", getExceptionName(ex));
-            builder.field("reason", ex.getMessage());
+            builder.field(TYPE, getExceptionName(ex));
+            builder.field(REASON, ex.getMessage());
             if (ex.getCause() != null) {
-                builder.field("caused_by");
+                builder.field(CAUSED_BY);
                 builder.startObject();
                 toXContent(builder, params, ex.getCause());
                 builder.endObject();
             }
             if (params.paramAsBoolean(REST_EXCEPTION_SKIP_STACK_TRACE, REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT) == false) {
-                builder.field("stack_trace", ExceptionsHelper.stackTrace(ex));
+                builder.field(STACK_TRACE, ExceptionsHelper.stackTrace(ex));
             }
         }
     }
 
     /**
+     * Generate a {@link ElasticsearchException} from a {@link XContentParser}. This does not
+     * returns the original exception type (ie NodeClosedException for example) but just wraps
+     * the type, the reason and the cause of the exception.
+     */
+    public static ElasticsearchException fromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token = ensureFieldName(parser, parser::nextToken);
+
+        String type = null, reason = null, stack = null;
+        ElasticsearchException cause = null;
+        Map<String, Object> headers = new HashMap<>();
+
+        String currentFieldName = parser.currentName();
```

Comment:
we never check that there is a start object, but we look for an end object. can we fix that? otherwise toXContent and fromXContent are not symmetric.